logic/src/commonMain/kotlin/com/wire/kalium/logic/data/user/UserRepository.kt

```diff
@@ -0,0 +1,54 @@
+package com.wire.kalium.logic.data.user
+
+import com.wire.kalium.logic.CoreFailure
+import com.wire.kalium.logic.functional.Either
+import com.wire.kalium.network.api.user.self.SelfApi
+import com.wire.kalium.network.utils.isSuccessful
+import com.wire.kalium.persistence.dao.MetadataDAO
+import com.wire.kalium.persistence.dao.QualifiedID
+import com.wire.kalium.persistence.dao.UserDAO
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.emptyFlow
+import kotlinx.coroutines.flow.map
+import kotlinx.serialization.decodeFromString
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+
+interface UserRepository {
+    suspend fun fetchSelfUser(): Either<CoreFailure, Unit>
+    suspend fun getSelfUser(): Flow<SelfUser?>
+}
+
+class UserDataSource(
+    private val userDAO: UserDAO,
+    private val metadataDAO: MetadataDAO,
+    private val selfApi: SelfApi,
+    private val userMapper: UserMapper): UserRepository {
+
+    override suspend fun fetchSelfUser(): Either<CoreFailure, Unit> {
+        val selfInfoResponse = selfApi.getSelfInfo()
+
+        return if (!selfInfoResponse.isSuccessful()) {
+            Either.Left(CoreFailure.ServerMiscommunication)
+        } else {
+            val user = userMapper.fromApiModelToDaoModel(selfInfoResponse.value)
+            userDAO.insertUser(user)
+            metadataDAO.insertValue(Json.encodeToString(user.id), SELF_USER_ID_KEY)
+            Either.Right(Unit)
+        }
+    }
+
+    override suspend fun getSelfUser(): Flow<SelfUser?> {
+        metadataDAO.valueByKey(SELF_USER_ID_KEY)?.let { encodedValue ->
+            val selfUserID: QualifiedID = Json.decodeFromString(encodedValue)
+            return userDAO.getUserByQualifiedID(selfUserID).map { it?.let { userMapper.fromDaoModel(it) } }
+        }
+
+        return emptyFlow()
```

Comment:
Hmm when I re-wrote this method like you suggested above with `filterNotNull()` it will return a non-optional `Flow<SelfUser>`.  Which I suspect will be an empty flow if there's no self user in DB.

Reply:
I'm not sure what SQLDelight does when there are no results for the query actually.

I assume that these Flows are never-ending. If there is no User in the databse, it will just stay open without emiting anything.

It's still a hidden failure (is it really ?) scenario where the `User` was not fetched, but it's being read 🤔. And the consumer is there waiting.

I am not sure we need to solve/improve this for now. But I am happy to keep thinking about this for the next iterations.